### PR TITLE
enable autoTP for MPT

### PR DESCRIPTION
**Pull request by sywangyi**

### CODE REVIEW

**Comment by sywangyi** [comment]
@yao-matrix @delock please help review

**Comment by sywangyi** [comment]
verified on model https://huggingface.co/mosaicml/mpt-7b https://huggingface.co/mosaicml/mpt-30b

**Comment by delock** [comment]
Hi @tjruwase is @molly-smith owner of AutoTP related changes?  There are contribution to enable AutoTP for different LLM model, should we invite Molly to review?

**Comment by delock** [comment]
@sywangyi  can you also add mpt into supported models in docs? 
 https://github.com/microsoft/DeepSpeed/blob/master/docs/_tutorials/automatic-tensor-parallelism.md

**Comment by tjruwase** [comment]
> Hi @tjruwase is @molly-smith owner of AutoTP related changes? There are contribution to enable AutoTP for different LLM model, should we invite Molly to review?

@delock, sure I will ask for Molly for help. Thanks!

**Comment by tjruwase** [comment]
> > Hi @tjruwase is @molly-smith owner of AutoTP related changes? There are contribution to enable AutoTP for different LLM model, should we invite Molly to review?
> 
> @delock, sure I will ask for Molly for help. Thanks!

@delock, July 4th is this week and so there might be some delay with reviews as team members take some time off. Apologies for the inconvenience.

**Comment by delock** [comment]
> > > Hi @tjruwase is @molly-smith owner of AutoTP related changes? There are contribution to enable AutoTP for different LLM model, should we invite Molly to review?
> > 
> > 
> > @delock, sure I will ask for Molly for help. Thanks!
> 
> @delock, July 4th is this week and so there might be some delay with reviews as team members take some time off. Apologies for the inconvenience.

Got it, thanks for reminding!

File: deepspeed/module_inject/replace_module.py
Comment (guoyejun) [inline deepspeed/module_inject/replace_module.py:377]: just curious do you mean [3, num_head, dim_per_head, hidden_size(=num_head*dim_per_head)] ? thanks

Comment (sywangyi) [reply deepspeed/module_inject/replace_module.py:377]: yes. I will fix the comment, thanks

Comment (inkcherry) [reply deepspeed/module_inject/replace_module.py]: hi, @sywangyi  I have sent a PR for fix different types of fusedqkv weight last week. Including this type. 
https://github.com/microsoft/DeepSpeed/pull/3844/files

Comment (sywangyi) [reply deepspeed/module_inject/replace_module.py]: yes, I could leverage some change in https://github.com/microsoft/DeepSpeed/pull/3844/files, will update PR once 3844 is merged.

**Comment by molly-smith** [comment]
@tjruwase @sywangyi @delock AutoTP creates a dictionary of which model layers require all reduce and is implemented in deepspeed/module_inject/auto_tp.py. This part is actually already functional for MPT. However, I assume the lower level implementation of tensor parallelism is not functional for MPT. I am not the code owner for this part of TP. I think @RezaYazdaniAminabadi is the right person to review these changes.

**Comment by tjruwase** [comment]
@molly-smith, thanks for the clarification. 

@RezaYazdaniAminabadi, can you please help review this?

File: deepspeed/module_inject/replace_module.py
Comment (RezaYazdaniAminabadi) [inline deepspeed/module_inject/replace_module.py:390]: may I ask that you use the `strided_copy` from `mp_replace` class (or add a new function) just to simplify the logic here?

File: deepspeed/inference/engine.py
Comment (RezaYazdaniAminabadi) [inline deepspeed/inference/engine.py:103]: that's a nice trick to make it MP-compatible. Thanks :) A minor comment here is to maybe move this function as well as the `build_alibi_...` to a a different file as utility for injection when auto-tp is used.

**Review by RezaYazdaniAminabadi — APPROVED** [review]

File: deepspeed/inference/engine.py
Comment (sywangyi) [reply deepspeed/inference/engine.py:103]: I do not know where I should place the function. Any suggestion?

File: deepspeed/inference/engine.py
Comment (mrwyattii) [inline deepspeed/inference/engine.py:107]: The import should be at the top of the file with the rest of the imports. Also, could we please move this to wherever @RezaYazdaniAminabadi suggests? Thanks!

**Review by mrwyattii — CHANGES_REQUESTED** [review]

File: deepspeed/inference/engine.py
Comment (sywangyi) [reply deepspeed/inference/engine.py:107]: @RezaYazdaniAminabadi should I place it to https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/module_inject/inject.py?

Comment (delock) [reply deepspeed/inference/engine.py]: Is it for model specific treatment of AutoTP?  I think a new file i.e. module_inject/auto_tp_model_utils.py to place all model specific policies might be a good place  @sywangyi  @RezaYazdaniAminabadi @mrwyattii

**Comment by sywangyi** [comment]
> Hi @sywangyi, I was trying out deepspeed for MPT30B as well and saw you had added the changes.
> 
> I just tried out your changes by building deepspeed from source by cloning the repository with changes and doing `pip install .`
> 
> ```
> # tried with both replace_with_kernel_inject as True and False
> ds_engine = deepspeed.init_inference(model, mp_size=4, dtype=torch.float16, replace_with_kernel_inject=True)
> model = ds_engine.module
> # evaluate model
> ```
> 
> I am getting the following error when replace_with_kernel_inject is False
> 
> ```
> │ a00f1a5307c5ca947e72e89693c/attention.py:41 in scaled_multihead_dot_product_attention            │
> │                                                                                                  │
> │    38 │   │   attn_bias = attn_bias[:, :, _s_q:, _s_k:]                                          │
> │    39 │   │   if attn_bias.size(-1) != 1 and attn_bias.size(-1) != s_k or (attn_bias.size(-2)    │
> │    40 │   │   │   raise RuntimeError(f'attn_bias (shape: {attn_bias.shape}) is expected to bro   │
> │ ❱  41 │   │   attn_weight = attn_weight + attn_bias                                              │
> │    42 │   min_val = torch.finfo(q.dtype).min                                                     │
> │    43 │   if key_padding_mask is not None:                                                       │
> │    44 │   │   if attn_bias is not None:                                                          │
> ╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
> RuntimeError: The size of tensor a (16) must match the size of tensor b (64) at non-singleton dimension 1
> ```
> 
> and w/o it I am getting OOM with 4xA100 40GB machines.
> 
> Anything I should be doing differently?

I could run mpt30B 4TP successfully in 4 RTX 8000 using AutoTP, seem the attn_bias is incorrect  in your side, could you have a check if https://github.com/microsoft/DeepSpeed/pull/3861/files#diff-27be33d45da8a29a59628046c212ecbdb630e85a9bd987a98431272f1472a3fbR63 is correctly called.

**Comment by minionKP** [comment]
> > Hi @sywangyi, I was trying out deepspeed for MPT30B as well and saw you had added the changes.
> > I just tried out your changes by building deepspeed from source by cloning the repository with changes and doing `pip install .`
> > ```
> > # tried with both replace_with_kernel_inject as True and False
> > ds_engine = deepspeed.init_inference(model, mp_size=4, dtype=torch.float16, replace_with_kernel_inject=True)
> > model = ds_engine.module
> > # evaluate model
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am getting the following error when replace_with_kernel_inject is False
> > ```
> > │ a00f1a5307c5ca947e72e89693c/attention.py:41 in scaled_multihead_dot_product_attention            │
> > │                                                                                                  │
> > │    38 │   │   attn_bias = attn_bias[:, :, _s_q:, _s_k:]                                          │
> > │    39 │   │   if attn_bias.size(-1) != 1 and attn_bias.size(-1) != s_k or (attn_bias.size(-2)    │
> > │    40 │   │   │   raise RuntimeError(f'attn_bias (shape: {attn_bias.shape}) is expected to bro   │
> > │ ❱  41 │   │   attn_weight = attn_weight + attn_bias                                              │
> > │    42 │   min_val = torch.finfo(q.dtype).min                                                     │
> > │    43 │   if key_padding_mask is not None:                                                       │
> > │    44 │   │   if attn_bias is not None:                                                          │
> > ╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
> > RuntimeError: The size of tensor a (16) must match the size of tensor b (64) at non-singleton dimension 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and w/o it I am getting OOM with 4xA100 40GB machines.
> > Anything I should be doing differently?
> 
> I could run mpt30B 4TP successfully in 4 RTX 8000 using AutoTP, seem the attn_bias is incorrect in your side, could you have a check if https://github.com/microsoft/DeepSpeed/pull/3861/files#diff-27be33d45da8a29a59628046c212ecbdb630e85a9bd987a98431272f1472a3fbR63 is correctly called.

I did a mistake in my setup with your changes. Can confirm it now works. Thanks

**Comment by sywangyi** [comment]
@mrwyattii I have remove the model specific ops to a specific file. could you review the change?

**Review by mrwyattii — APPROVED** [review]

**Pull request merged**
